content/recipes/authoring/kata-snippets/submission-tests.md

```diff
@@ -0,0 +1,88 @@
+  - Extra time spent on running a reference solution
+  - Reference solution being accessible to users by mistake
+  - Input mutation by the user solution which can affect the input passed to the reference solution, or make assertion messages confusing
+  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
+- **The reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
+- **The reference solution shouldn't be accessible to the user solution.** It should not be possible to call the reference solution directly, or implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function. Check the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs and outputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every aspect of the specification in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s). Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
+- **Tests should check the solution with edge cases and cases that require special handling.** For example, empty arrays and empty strings should be covered by fixed tests unless the kata description _explicitly_ states that such inputs need not be considered.
+
+
+## Random Tests
+
+Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.
+
+- **Keep the amount of random tests reasonable.** If random tests are not used to verify performance, keep their amount small enough to guarantee sufficient coverage. In some cases it's not necessary to have dozens, or even hundreds of random tests, and several of them would be enough. Do not create random tests just for the sake of their amount. However, if you believe that a large amount of random tests helps you to detect invalid solutions which are not caught by fixed tests, you should try and find a pattern in what types of cases fail particularly often and extract them to the corresponding fixed tests.
```

Comment:
```suggestion
- **Keep the amount of random tests reasonable.** If random tests are not used to verify performances, keep their amount small but big enough to guarantee sufficient coverage. In some cases it's not necessary to have dozens, or even hundreds of random tests, and several of them would be enough. Do not create random tests just for the sake of their amount. However, if you believe that a large amount of random tests helps you to detect invalid solutions which are not caught by fixed tests, you should try to find a pattern in what types of cases are showing up particularly often, extract them and add them to the fixed tests.
```

>  In some cases it's not necessary to have dozens, or even hundreds of random tests, and several of them would be enough

I tend to disagree on putting the sentence above here. It's perfectly correct (I bet it's coming from unnamed?), but most readers will read this as "no need to put a lot of them anyway", even when they do not understand what their random tests are actually generating or not. So I'd let that as a side note or I'd just not put it there. Better having users putting systematically 100 random tests even when useless than the opposite, imo.

Reply:
It's related to [this comment](https://github.com/codewars/docs/issues/19#issuecomment-723370170) from discussion thread, and it's meant to explicitly drive users away from spamming random tests in hope they will generate all possible kinds of input. It's explicitly complemented by the sentence _"[random tests] should generate test cases for all scenarios which cannot be completely tested with fixed tests."_ below, so authors put more effort into discovering possible scenarios rather than just spinning 1000 times in hope to catch _something_ .

Main idea here is: if 10 tests give just as good coverage as 100, then 5 will be enough.

Reply:
OK, makes sense. But that's not exactly what I got from the sentences.
What about his?

```suggestion
- **Keep the amount of random tests reasonable if random tests are not used to verify performances.** Keep their amount as small as possible as long as a good coverage is still guaranteed: in some situations, only a bunch of inputs are actually testable, so no need to tests 10 times each of them, so use randomized testing instead. Otherwise 100 random tests are generally enough, or maybe less, depending on the task/situation.
- **Don't hesitate to build different kind of random input generators**. If a specific kind of input has very little chances to occur purely at random (like... generating a palindrom) it's better to build a specific random generator that can enforce this kind of input rather than rely on 1000 random tests and just pray for the edge case to come up (see point above). It happens that a set several random generators are used to fullfil the different needs. Note: keep in mind those generators still need to be random enough, and keeping one fully random generator is alwys good (it may generate cases you didn't think about).
```

well, it's way longer, with one new point, but... x)

Reply:
Fixed